Target-MVD-Mathias/AppDelegate.swift

```diff
@@ -42,13 +44,79 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
       self.window?.rootViewController = vc
     }
     
+    PushNotificationManager.push().delegate = self
+    
+    // set default Pushwoosh delegate for iOS10 foreground push handling
+    if #available(iOS 10.0, *) {
+      UNUserNotificationCenter.current().delegate = PushNotificationManager.push().notificationCenterDelegate
+    }
+    
+    // track application open statistics
+    PushNotificationManager.push().sendAppOpen()
+    
     return true
   }
   
   func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {
     return FBSDKApplicationDelegate.sharedInstance().application(application, open: url, sourceApplication: sourceApplication, annotation: annotation)
   }
   
+  func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+    PushNotificationManager.push().handlePushRegistration(deviceToken as Data!)
+    
+    if let token = PushNotificationManager.push().getPushToken() {
+      UserAPI.updatePushToken(token: token)
+    }
+  }
+  
+  func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+    PushNotificationManager.push().handlePushRegistrationFailure(error)
+  }
+  
+  func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any],
+                   fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
+    PushNotificationManager.push().handlePushReceived(userInfo)
+    completionHandler(UIBackgroundFetchResult.noData)
+  }
+  
+  func onPushAccepted(_ pushManager: PushNotificationManager!, withNotification pushNotification: [AnyHashable : Any]!, onStart: Bool) {
+    if let pushBody = pushNotification["u"] as? String, let data = pushBody.data(using: .utf8) {
+      do {
+        guard let body = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
```

Comment:
you can make this simpler with `SwiftyJSON` since you are already using it later.